### ISSUE TITLE
Daemon process crashed due to max out memory

### ISSUE THREAD

**Issue by William8Work** [filecoin-project/lotus#7827]
### Checklist

- [X] This is **not** a security-related bug/issue. If it is, please follow please follow the [security policy](https://github.com/filecoin-project/lotus/security/policy).
- [X] This is **not** a question or a support request. If you have any lotus related questions, please ask in the [lotus forum](https://github.com/filecoin-project/lotus/discussions).
- [X] This is **not** a new feature request. If it is, please file a [feature request](https://github.com/filecoin-project/lotus/issues/new?assignees=&labels=need%2Ftriage%2Ckind%2Ffeature&template=feature_request.yml) instead.
- [X] This is **not** an enhancement request. If it is, please file a [improvement suggestion](https://github.com/filecoin-project/lotus/issues/new?assignees=&labels=need%2Ftriage%2Ckind%2Fenhancement&template=enhancement.yml) instead.
- [X] I **have** searched on the [issue tracker](https://github.com/filecoin-project/lotus/issues) and the [lotus forum](https://github.com/filecoin-project/lotus/discussions), and there is no existing related issue or discussion.
- [X] I am running the [`Latest release`](https://github.com/filecoin-project/lotus/releases), or the most recent RC(release canadiate) for the upcoming release or the dev branch(master), or have an issue updating to any of these.
- [X] I did not make any code changes to lotus.

### Lotus component

- [X] lotus daemon - chain sync
- [ ] lotus miner - mining and block production
- [ ] lotus miner/worker - sealing
- [ ] lotus miner - proving(WindowPoSt)
- [ ] lotus miner/market - storage deal
- [ ] lotus miner/market - retrieval deal
- [ ] lotus miner/market - data transfer
- [ ] lotus client
- [ ] lotus JSON-RPC API
- [ ] lotus message management (mpool)
- [ ] Other

### Lotus Version

```text
Daemon:  1.13.2-rc1+mainnet+git.52c7fa242+api1.2.0
Local: lotus-miner version 1.13.2-rc1+mainnet+git.52c7fa242
```

### Describe the Bug

After upgraded to v1.13.2-rc1 for two weeks, the lotus daemon where running fine. Today, it crashed due to maxing out memory.

This is a DELL PowerEdge Server machine running with Intel CPU, and 256GB RAM. It is a dedicated daemon machine with 
`/usr/local/bin/lotus daemon --config /media/cache/daemon/config.toml`

There is no lotus-miner, market, or lotus workers running on this machine. 

The deamon is using Splitstore. Here are the detail config.toml

```
[API]
  # Binding address for the Lotus API
  #
  # type: string
  ListenAddress = "/ip4/0.0.0.0/tcp/1234/http"

  # type: string
  RemoteListenAddress = "10.1.18.167:1234"

  # type: Duration
  #Timeout = "30s"


[Backup]
  # Note that in case of metadata corruption it might be much harder to recover
  # your node if metadata log is disabled
  #
  # type: bool
  #DisableMetadataLog = false


[Libp2p]
  # Binding address for the libp2p host - 0 means random port.
  # Format: multiaddress; see https://multiformats.io/multiaddr/
  #
  # type: []string
  #ListenAddresses = ["/ip4/0.0.0.0/tcp/0", "/ip6/::/tcp/0"]

  # Addresses to explicitally announce to other peers. If not specified,
  # all interface addresses are announced
  # Format: multiaddress
  #
  # type: []string
  #AnnounceAddresses = []

  # Addresses to not announce
  # Format: multiaddress
  #
  # type: []string
  #NoAnnounceAddresses = []

  # type: uint
  ConnMgrLow = 500

  # type: uint
  ConnMgrHigh = 1500

  # type: Duration
  ConnMgrGrace = "30s"


[Pubsub]
  # Run the node in bootstrap-node mode
  #
  # type: bool
  #Bootstrapper = false

  # type: string
  #RemoteTracer = ""


[Client]
  # type: bool
  #UseIpfs = false

  # type: bool
  #IpfsOnlineMode = false

  # type: string
  #IpfsMAddr = ""

  # type: bool
  #IpfsUseForRetrieval = false

  # The maximum number of simultaneous data transfers between the client
  # and storage providers
  #
  # type: uint64
  #SimultaneousTransfers = 20


[Wallet]
  # type: string
  #RemoteBackend = ""

  # type: bool
  #EnableLedger = false

  # type: bool
  #DisableLocal = false


[Fees]
  # type: types.FIL
  DefaultMaxFee = "0.10 FIL"


[Chainstore]
  # type: bool
  EnableSplitstore = true

  [Chainstore.Splitstore]
    # ColdStoreType specifies the type of the coldstore.
    # It can be "universal" (default) or "discard" for discarding cold blocks.
    #
    # type: string
    ColdStoreType = "discard"

    # HotStoreType specifies the type of the hotstore.
    # Only currently supported value is "badger".
    #
    # type: string
    #HotStoreType = "badger"

    # MarkSetType specifies the type of the markset.
    # It can be "map" (default) for in memory marking or "badger" for on-disk marking.
    #
    # type: string
    #MarkSetType = "map"

    # HotStoreMessageRetention specifies the retention policy for messages, in finalities beyond
    # the compaction boundary; default is 0.
    #
    # type: uint64
    #HotStoreMessageRetention = 0

    # HotStoreFullGCFrequency specifies how often to perform a full (moving) GC on the hotstore.
    # A value of 0 disables, while a value 1 will do full GC in every compaction.
    # Default is 20 (about once a week).
    #
    # type: uint64
    HotStoreFullGCFrequency = 50
```

### Logging Information

```text
logs from the system using dmesg:

2021-12-18T07:20:56,472613-08:00 Tasks state (memory values in pages):
2021-12-18T07:20:56,472614-08:00 [  pid  ]   uid  tgid total_vm      rss pgtables_bytes swapents oom_score_adj name
2021-12-18T07:20:56,472618-08:00 [   1525]  1000  1525 91062796 16513542 580308992        0             0 lotus
2021-12-18T07:20:56,472623-08:00 oom-kill:constraint=CONSTRAINT_MEMCG,nodemask=(null),cpuset=/,mems_allowed=0-1,oom_memcg=/system.slice/lotusdaemon.service,task_memcg=/system.slice/lotusdaemon.service,task=lotus,pid=1525,uid=1000
2021-12-18T07:20:56,472887-08:00 Memory cgroup out of memory: Killed process 1525 (lotus) total-vm:364251184kB, anon-rss:66054168kB, file-rss:0kB, shmem-rss:0kB, UID:1000 pgtables:566708kB oom_score_adj:0
2021-12-18T07:21:02,179487-08:00 oom_reaper: reaped process 1525 (lotus), now anon-rss:0kB, file-rss:4kB, shmem-rss:0kB
```

### Repo Steps

1. Run '...'
2. Do '...'
3. See error '...'
...

**Comment by William8Work** [comment]
Looks like this might be related to the split store compaction. It generates some errors right before the daemon was crashed:
```

2021-12-18T07:20:54.447-0800    INFO    badgerbs        v2@v2.2007.2/db.go:1511 Level: 5.      0 B Size.   2.7 TB Max.
2021-12-18T07:20:54.447-0800    INFO    badgerbs        v2@v2.2007.2/db.go:1511 Level: 6.      0 B Size.    27 TB Max.
2021-12-18T07:20:54.447-0800    INFO    badgerbs        v2@v2.2007.2/db.go:1475 Attempting to compact with {level:1 score:1.71 dropPrefixes:[]}
2021-12-18T07:20:54.447-0800    INFO    badgerbs        v2@v2.2007.2/levels.go:1000     [Compactor: 175] Running compaction: {level:1 score:1.71 dropPrefixes:[]} for level: 1
2021-12-18T07:20:54.447-0800    WARN    badgerbs        v2@v2.2007.2/db.go:1488 While running doCompact with {level:1 score:1.71 dropPrefixes:[]}. Error: Unable to fill tables
2021-12-18T07:20:54.448-0800    WARN    badgerbs        v2@v2.2007.2/db.go:1488 While running doCompact with {level:1 score:1.71 dropPrefixes:[]}. Error: Unable to fill tables
2021-12-18T07:20:54.448-0800    WARN    badgerbs        v2@v2.2007.2/db.go:1488 While running doCompact with {level:1 score:1.71 dropPrefixes:[]}. Error: Unable to fill tables
2021-12-18T07:20:54.448-0800    WARN    badgerbs        v2@v2.2007.2/db.go:1488 While running doCompact with {level:1 score:1.71 dropPrefixes:[]}. Error: Unable to fill tables
2021-12-18T07:20:54.448-0800    WARN    badgerbs        v2@v2.2007.2/db.go:1488 While running doCompact with {level:1 score:1.71 dropPrefixes:[]}. Error: Unable to fill tables
2021-12-18T07:20:54.448-0800    WARN    badgerbs        v2@v2.2007.2/db.go:1488 While running doCompact with {level:1 score:1.71 dropPrefixes:[]}. Error: Unable to fill tables
```

Attached is the details of log.

[f08399.log](https://github.com/filecoin-project/lotus/files/7739908/f08399.log)

**Comment by William8Work** [comment]
Daemon crashed again last night. This time it happened right before the window post so my miner last all its power.

I didn't see the splitstore activities before the crash timestamp.

**Comment by William8Work** [comment]
this time is v1.32.2-rc4, I upgraded right after the crash yesterday.

```
$ lotus version && lotus-miner version
Daemon:  1.13.2-rc4+mainnet+git.e05fdf892+api1.4.0
Local: lotus version 1.13.2-rc4+mainnet+git.e05fdf892
Daemon:  1.13.2-rc4+mainnet+git.e05fdf892+api1.2.0
Local: lotus-miner version 1.13.2-rc4+mainnet+git.e05fdf892
```

**Comment by William8Work** [comment]
```
the hardware and os are:
- Dell PowerEdge R720, CPU Xeon E5-2670 x2, 256 GB RAM
- 10GiB Network
- Running Daemon process only. Other lotus processes are on separate machines.
- Daemon runs on 1TiB nvme drive with 498GiB free. Free space change time to time due to the split store size

- Ubuntu 20.04 with latest updates
```

**Comment by scaseye** [comment]
lotus version 1.13.1+mainnet+git.8943de27e

Running lotus, markets, miner on one machine. Running pc1 and pc2/c2 workers on other nodes.

here is what was happening at the time. 
![Sealing jobs All (Running VS Idle)](https://user-images.githubusercontent.com/69066370/147262693-9fffce23-42d1-4349-8cb4-c22e890627b0.png)


```
$ sar -r
Linux 5.4.0-89-generic (homemate-server01) 	12/23/2021 	_x86_64_	(256 CPU)
12:00:01 AM kbmemfree   kbavail kbmemused  %memused kbbuffers  kbcached  kbcommit   %commit  kbactive   kbinact   kbdirty
05:55:01 AM 899012124 1501485104 585931656     27.89   1631908 595928144 631303636     30.04 871189328 307428800   7114264
06:05:01 AM 913265740 1540437420 546964532     26.03   1637500 616201448 631688872     30.06 865943256 294069968      4692
06:15:01 AM 894711780 1559909060 527490956     25.10   1639272 655831564 632156380     30.08 879868080 300294332       772
06:25:01 AM 884155916 1526249696 561285740     26.71   1643020 634491344 627993304     29.89 915478972 277397216      4924
```

```
Dec 23 06:17:15 homemate-server01 kernel: lotus invoked oom-killer: gfp_mask=0xcc0(GFP_KERNEL), order=0, oom_score_adj=0
Dec 23 06:17:15 homemate-server01 kernel: CPU: 78 PID: 1336761 Comm: lotus Tainted: P           OE     5.4.0-89-generic #100-Ubuntu
Dec 23 06:17:15 homemate-server01 kernel: Hardware name: GIGABYTE G492-Z51-00/MZ52-G40-00, BIOS R17 02/25/2021
Dec 23 06:17:15 homemate-server01 kernel: Call Trace:
Dec 23 06:17:15 homemate-server01 kernel:  dump_stack+0x6d/0x8b
Dec 23 06:17:15 homemate-server01 kernel:  dump_header+0x4f/0x1eb
Dec 23 06:17:15 homemate-server01 kernel:  oom_kill_process.cold+0xb/0x10
Dec 23 06:17:15 homemate-server01 kernel:  out_of_memory.part.0+0x1df/0x3d0
Dec 23 06:17:15 homemate-server01 kernel:  out_of_memory+0x6d/0xd0
Dec 23 06:17:15 homemate-server01 kernel:  mem_cgroup_out_of_memory+0xbd/0xe0
Dec 23 06:17:15 homemate-server01 kernel:  try_charge+0x77c/0x810
Dec 23 06:17:15 homemate-server01 kernel:  ? __alloc_pages_nodemask+0x173/0x320
Dec 23 06:17:15 homemate-server01 kernel:  mem_cgroup_try_charge+0x71/0x190
Dec 23 06:17:15 homemate-server01 kernel:  mem_cgroup_try_charge_delay+0x22/0x50
Dec 23 06:17:15 homemate-server01 kernel:  do_anonymous_page+0x13c/0x650
Dec 23 06:17:15 homemate-server01 kernel:  __handle_mm_fault+0x760/0x7a0
Dec 23 06:17:15 homemate-server01 kernel:  handle_mm_fault+0xca/0x200
Dec 23 06:17:15 homemate-server01 kernel:  do_user_addr_fault+0x1f9/0x450
Dec 23 06:17:15 homemate-server01 kernel:  __do_page_fault+0x58/0x90
Dec 23 06:17:15 homemate-server01 kernel:  do_page_fault+0x2c/0xe0
Dec 23 06:17:15 homemate-server01 kernel:  page_fault+0x34/0x40
Dec 23 06:17:15 homemate-server01 kernel: RIP: 0033:0x5a55b3
Dec 23 06:17:15 homemate-server01 kernel: Code: 00 48 81 eb 00 01 00 00 48 81 c7 00 01 00 00 48 81 fb 00 01 00 00 73 82 e9 07 ff ff ff c5 fd ef c0 48 81 fb 00 00 00 02 73 46 <c5> fe 7f 07 c5 fe 7f 47 20 c5 fe 7f 47 40 c5 fe 7f 47 60 48 81 eb
Dec 23 06:17:15 homemate-server01 kernel: RSP: 002b:000000c05fe80d98 EFLAGS: 00010202
Dec 23 06:17:15 homemate-server01 kernel: RAX: 0000000000000000 RBX: 0000000000278000 RCX: 000000c110e12000
Dec 23 06:17:15 homemate-server01 kernel: RDX: 0000000000080e00 RSI: 0000000000000002 RDI: 000000c111b9c000
Dec 23 06:17:15 homemate-server01 kernel: RBP: 000000c05fe80de0 R08: 000000c110e12000 R09: 00007f9ab0306070
Dec 23 06:17:15 homemate-server01 kernel: R10: 0000000000000801 R11: 0000000000000f09 R12: 0000000000000800
Dec 23 06:17:15 homemate-server01 kernel: R13: 000000000000000f R14: 000080c110e12000 R15: 000080c111e13fff
Dec 23 06:17:15 homemate-server01 kernel: memory: usage 10485756kB, limit 10485760kB, failcnt 15367657
Dec 23 06:17:15 homemate-server01 kernel: memory+swap: usage 0kB, limit 9007199254740988kB, failcnt 0
Dec 23 06:17:15 homemate-server01 kernel: kmem: usage 461292kB, limit 9007199254740988kB, failcnt 0
Dec 23 06:17:15 homemate-server01 kernel: Memory cgroup stats for /system.slice/lotus-daemon.service:
Dec 23 06:17:15 homemate-server01 kernel: anon 10253357056
                                          file 0
                                          kernel_stack 8110080
                                          slab 164147200
                                          sock 0
                                          shmem 0
                                          file_mapped 0
                                          file_dirty 17166336
                                          file_writeback 0
                                          anon_thp 1795162112
                                          inactive_anon 0
                                          active_anon 10252787712
                                          inactive_file 0
                                          active_file 0
                                          unevictable 0
                                          slab_reclaimable 69636096
                                          slab_unreclaimable 94511104
                                          pgfault 65104083
                                          pgmajfault 5324616
                                          workingset_refault 62496489
                                          workingset_activate 13895508
                                          workingset_nodereclaim 0
                                          pgrefill 70548707
                                          pgscan 99493251
                                          pgsteal 65262732
                                          pgactivate 34224036
                                          pgdeactivate 47943850
                                          pglazyfree 0
                                          pglazyfreed 0
                                          thp_fault_alloc 1287
                                          thp_collapse_alloc 0
Dec 23 06:17:15 homemate-server01 kernel: Tasks state (memory values in pages):
Dec 23 06:17:15 homemate-server01 kernel: [  pid  ]   uid  tgid total_vm      rss pgtables_bytes swapents oom_score_adj name
Dec 23 06:17:15 homemate-server01 kernel: [1331950]  1000 1331950 67055412  3946742 278962176        0             0 lotus
Dec 23 06:17:15 homemate-server01 kernel: oom-kill:constraint=CONSTRAINT_MEMCG,nodemask=(null),cpuset=/,mems_allowed=0-1,oom_memcg=/system.slice/lotus-daemon.service,task_memcg=/system.slice/lotus-daemon.service,task=lotus,pid=1331950,uid=1000
Dec 23 06:17:15 homemate-server01 kernel: Memory cgroup out of memory: Killed process 1331950 (lotus) total-vm:268221648kB, anon-rss:10023044kB, file-rss:5763924kB, shmem-rss:0kB, UID:1000 pgtables:272424kB oom_score_adj:0
Dec 23 06:17:15 homemate-server01 kernel: oom_reaper: reaped process 1331950 (lotus), now anon-rss:0kB, file-rss:5772600kB, shmem-rss:0kB

```

Service File + Environment file.
```
cat /etc/systemd/system/lotus-daemon.service 
[Unit]
Description=Lotus Daemon
After=network-online.target
Requires=network-online.target

[Service]
Environment=GOLOG_FILE='/home/scaseye/logs/daemon.log'
Environment=GOLOG_LOG_FMT='json'
EnvironmentFile=/home/scaseye/env/common.env
ExecStart=/usr/local/bin/lotus daemon
Restart=always
RestartSec=10
User=scaseye
Group=scaseye

MemoryAccounting=true
MemoryHigh=8G
MemoryMax=10G
#LimitNOFILE=8192:10240
LimitNOFILE=32000000:128000000

[Install]
WantedBy=multi-user.target
```

```
#############################
# cat /home/scaseye/env/common.env
# TOKENS
FULLNODE_API_INFO=[token]:/ip4/40.0.0.199/tcp/1234/http
MINER_API_INFO=[token]:/ip4/40.0.0.199/tcp/2345/http
MARKETS_API_INFO=[token]:/ip4/40.0.0.199/tcp/8787/http
#PATH
LOTUS_MARKETS_PATH=/filecoin/mainnet32GiB/.lotusmarkets
LOTUS_MINER_PATH=/filecoin/mainnet32GiB/.lotusminer
LOTUS_PATH=/filecoin/mainnet32GiB/.lotus
FIL_PROOFS_PARAMETER_CACHE=/home/scaseye/.lotusproofs
FIL_PROOFS_PARENT_CACHE=/home/scaseye/.lotusproofs
TMPDIR=/filecoin/mainnet32GiB/.lotustmpgpu
#Config/build/Performance
FFI_USE_CUDA=1
BELLMAN_CUSTOM_GPU="NVIDIA RTX A6000:10752"
BELLMAN_CPU_UTILIZATION=0.875
RUST_BACKTRACE=full
RUST_LOG=debug
RUSTFLAGS="-C target-cpu=native -g"
FFI_BUILD_FROM_SOURCE=1
FIL_PROOFS_USE_GPU_COLUMN_BUILDER=1
FIL_PROOFS_USE_GPU_TREE_BUILDER=1
FIL_PROOFS_MAXIMIZE_CACHING=1
FIL_PROOFS_USE_MULTICORE_SDR=1
FIL_PROOFS_VERIFY_CACHE=1

```

**Comment by scaseye** [comment]
im going to try these settings in the daemon service and see what happens.

```
MemoryAccounting=true
MemoryHigh=56G
MemoryMax=64G
LimitNOFILE=18192:102140
```

**Comment by William8Work** [comment]
Thank you @scaseye.

Additional updates: 

The lotus daemon process crashed twice during this few days and one of them caused my power lost. After restart, I turned on the memory monitoring, and removed the memory control. I saw the following graph:

![Screen Shot 2021-12-19 at 7 57 52 PM](https://user-images.githubusercontent.com/3475205/147429928-4d276210-f6fe-48c2-81d2-ca4c78df173e.png)

I shut it down and immediate restarted to clear off the memory. After that, it has been running fine without crashing last 5 days. And the memory was mostly at 20-30GiB range.
